### PR TITLE
EDU-267 Render media in markdown

### PR DESCRIPTION
**Pull request by ahelmberger**
Closes https://educandu.atlassian.net/browse/EDU-267

### CODE REVIEW

File: migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
+// eslint-disable-next-line camelcase
+export default class Educandu_2022_01_21_01_add_render_media_to_markdown_and_annotation {
```
Comment (adrianaluchian) [inline migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js:2]: You copied the `initial-schema` script and then transformed the original into this current script and now there's a `educandu-2022-01-17-01-initial-schema copy.js` in place of the original 😬 

I've done this myself a couple of times 😅

File: src/utils/string-utils.js [hunk 2]
```diff
@@ -1,19 +1,18 @@
...
+  '<': '&lt;',
+  '>': '&gt;',
+  '"': '&quot;'
 };
+
+const HTML_ESCAPE_TEST_PATTERN = new RegExp(`[${Object.keys(HTML_REPLACEMENT_MAP).map(escapeStringRegexp).join('')}]`);
+
+const HTML_ESCAPE_REPLACEMENT_PATTERN = new RegExp(HTML_ESCAPE_TEST_PATTERN.source, 'g');
+
+export function escapeHtml(str) {
```
Comment (adrianaluchian) [inline src/utils/string-utils.js:14]: Maybe we should then use this method also in the `page-renderer` and `mail-service` and remove the external `htmlescape` dependency 🤔 What do you think?

File: migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js (hunk 1 above)
Comment (ahelmberger) [reply migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js:2]: :scream: I should check the changed files more carefully, fixed

File: migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    const updateFunc = section => {
+      delete section.content.renderMedia;
+    };
+
+    await this.updateSectionsInCollection('documentRevisions', ['markdown', 'annotation'], updateFunc);
+    await this.updateSectionsInCollection('documents', ['markdown', 'annotation'], updateFunc);
+  }
+
+  async updateSectionsInCollection(collectionName, sectionTypes, updateFunc) {
+    const filter = { $or: sectionTypes.map(st => ({ 'sections.type': st })) };
```
Comment (adrianaluchian) [inline migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js:25]: You could also filter more simply by `{ 'sections.type': { $in: sectionTypes } }`

File: src/utils/string-utils.js (hunk 2 above)
Comment (ahelmberger) [reply src/utils/string-utils.js:14]: They don't do exactly the same though. https://www.npmjs.com/package/htmlescape escapes JSON for usage as an object literal inside of a `<script>` tag. They escape slightly different characters.

File: migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    await this.updateAll(collectionName, filter, doc => {
+      const hasUpdates = doc.sections.reduce((accu, section) => accu || this.tryUpdateSection(section, sectionTypes, updateFunc), false);
+      // eslint-disable-next-line no-console
+      console.log(`Collection: ${collectionName}, ID: ${doc._id} -> ${hasUpdates ? 'updated' : 'not modified'}`);
```
Comment (adrianaluchian) [inline migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js:29]: You could maybe put this log after line 55 and only log the relevant information, I think "not modified" might just pollute the logs.

File: migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js (hunk 3 above)
Comment (ahelmberger) [reply migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js:25]: Of course, thanks, I didn't think of it. Changed

File: migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js (hunk 4 above)
Comment (ahelmberger) [reply migrations/educandu-2022-01-21-01-add-render-media-to-markdown-and-annotation.js:29]: Yes, changed

File: src/common/github-flavored-markdown.js [hunk 5]
```diff
@@ -1,42 +1,82 @@
...
+function educanduFeatures(md) {
+  md.core.ruler.after('inline', 'educandu-features', state => {
+    const { cdnRootUrl, renderMedia, collectCdnUrl } = state.env;
+
+    state.tokens.filter(parentToken => parentToken.type === 'inline').forEach(parentToken => {
+      parentToken.children?.forEach(token => {
+        if (token.type === 'link_open' || token.type === 'image') {
+          let targetUrl = null;
+          const attrToReplace = token.type === 'link_open' ? 'href' : 'src';
+          token.attrs.forEach(attr => {
```
Comment (adrianaluchian) [inline src/common/github-flavored-markdown.js:26]: Could `token.attrs.forEach` be replaced with a `token.attrs.find(attr => attr[0] === attrToReplace);`?
Or maybe I got this code wrong, super hard to follow 😳

Comment (ahelmberger) [reply src/common/github-flavored-markdown.js]: True, it saves one level of nesting, done, thanks!

Comment (ahelmberger) [reply src/common/github-flavored-markdown.js]: I could boil it down using the functions provided by the node class in markdown-it, so I think it looks much more understandable now. Thanks for the input!

Comment (adrianaluchian) [reply src/common/github-flavored-markdown.js]: Looks great, thank you, now the code speaks for itself! 🙏

**Review by adrianaluchian — APPROVED** [review]
